### ISSUE TITLE
`tmt test show` verbosity + adjust rules

### ISSUE THREAD

**Issue by lukaszachy** [teemtee/tmt#1813]
Original use case: As a user I want to see which adjust rules are defined for the node (test, story or plan) so I can make sure node metadata are evaluated properly.

During the discussion we specified verbosity levels as following. Note that verbosity us for user so it should affect output user needs to see. Debug level is indented for tmt developer or more advanced user.

Verbosity levels:
  0 … non-default attributes (enabled true is hidden)
  1 … all defined attributes (including defaults)
  2 … fmf-id, web link, applied adjust rules
  3 … original metadata (including extra-*), all adjust rules

To be decided:
##### What if adjust changes value back to the default value, should it be printed in verbosity 0?

 E.g.  `tmt -c distro=fedora test show`  for the test bellow should print "enabled: true"?
```
enabled: false
adjust:
  - when: distro == fedora
    enabled: true
```

##### How to visualize explicit empty values? 

There is difference between not defined (e.g. require missing in user metadata) and set to empty (e..g package is installed by default)
```
require: []
```

**Comment by psss** [comment]
> What if adjust changes value back to the default value, should it be printed in verbosity 0?

I'd say it shouldn't be displayed: For given context the `enabled` key has the default value so as a user who wants to quickly check the basic information about the test I'm not interested in this detail.

> How to visualize explicit empty values?

I'd go with the empty list for such cases. The important part is the final set of the values. If interested in the original content of the `fmf` file user should use `-vvv`.

**Comment by happz** [comment]
> > What if adjust changes value back to the default value, should it be printed in verbosity 0?
> 
> I'd say it shouldn't be displayed: For given context the `enabled` key has the default value so as a user who wants to quickly check the basic information about the test I'm not interested in this detail.

It might be also hard to realize the value has changed. There are helpers to extract default values, and one can compare it with the actual one, but if they are the same...

> 
> > How to visualize explicit empty values?
> 
> I'd go with the empty list for such cases. The important part is the final set of the values. If interested in the original content of the `fmf` file user should use `-vvv`.

> 3 … original metadata (including extra-*), all adjust rules

IIUIC, tmt's `show()` would, with `-vvv`, ignore object attributes, validated & normalized, and instead show the content of the raw fmf node? The good thing is, such behavior would be in one place only, the bad one might be slightly complex code needed to deal with possibly invalid content of such a node. Invalid keys might be present, mandatory keys might be missing, just something to be aware of.

**Comment by lukaszachy** [comment]
> IIUIC, tmt's `show()` would, with `-vvv`, ignore object attributes, validated & normalized, and instead show the content of the raw fmf node?

Not really. IMO it would be very confusing if values between -vv and -vvv changed. As I user expect _more_ information to be printed, not different one. 

IIRC the  `extra-*` are treated as a free-form key-value to store whatever user wants but doesn't expect to be recognized by tmt (export/import writes extra-task,-summary and -nitrate though) so we can print them as if they were "official".

As for the original values: How about having new key 'original' (name among those lines) which would contain the raw fmf node? (as in whatever is in the node gets printed, no normalization).


Lets say I have fmf file:
```yaml
test: echo
adjust:
- when: distro == fedora
  test+: " hello fedora"
- when: distro == rhel
  test+: " rhel"
extra-summary: lorem ipsum
tagZ: [] -- intentional typo in the key name
```

```
$ tmt -c distro=fedora test show
/ 
  test: echo hello fedora

$ tmt -c distro=fedora test show -v
/ 
  test: echo hello fedora
  enabled: true
  framework: shell
  ...

$ tmt -c distro=fedora test show -vv
/ 
  test: echo hello fedora
  enabled: true
  framework: shell
  ...
  fmf-id
  web
  adjusted by: 
  - when: distro == fedora
     test+: " hello fedora"

$ tmt -c distro=fedora test show -vvv
/ 
  test: echo hello fedora
  enabled: true
  framework: shell
  ...
  fmf-id
  web
  adjusted by: 
  - when: distro == fedora
     test+: " hello fedora"
  adjust:
     - when: distro == fedora
       test+: " hello fedora"
    - when: distro == rhel
      test+: " rhel"
  extra-summary: lorem ipsum
  raw_fmf_data: 
    test: echo
    adjust:
    - when: distro == fedora
      test+: " hello fedora"
    - when: distro == rhel
      test+: " rhel"
    extra-summary: lorem ipsum
    tagZ: [] -- intentional typo in the key name
```

I'm not sure from which level `require: []` should appear. Test doesn't define it so maybe in `-vv`? There is no default value (I'd argue that setting it to [] is just to make our life easier without None than the default value)

**Comment by happz** [comment]
> > IIUIC, tmt's `show()` would, with `-vvv`, ignore object attributes, validated & normalized, and instead show the content of the raw fmf node?
> 
> Not really. IMO it would be very confusing if values between -vv and -vvv changed. As I user expect _more_ information to be printed, not different one.

+1

> 
> IIRC the `extra-*` are treated as a free-form key-value to store whatever user wants but doesn't expect to be recognized by tmt (export/import writes extra-task,-summary and -nitrate though) so we can print them as if they were "official".
> 
> As for the original values: How about having new key 'original' (name among those lines) which would contain the raw fmf node? (as in whatever is in the node gets printed, no normalization).

That sounds like the best of both worlds. Raw node data are still available, and `self.node` should be always present, showing it under a dedicated key looks like a good setup to me.

> I'm not sure from which level `require: []` should appear. Test doesn't define it so maybe in `-vv`? There is no default value (I'd argue that setting it to [] is just to make our life easier without None than the default value)

I always considered `[]` to be the default value, as in "the key is a *list* of requirements", then the default in this sense would be an empty list. The fact it removes the need for `Optional` and "for package do yum install" iterations are then simply no-ops is a very welcome bonus.

At which level to print it, `[]` would be equal to the default value, so it seems "1 … all defined attributes (including defaults)" would apply. With helpers like `listed` it doesn't have to look that ugly :)

**Comment by psss** [comment]
Seems there is quite a consensus about the approach, minor improvement, not blocking, adding the backlog as a nice-to-have.
